### ISSUE TITLE
Create a cantilever beam topology optimization problem.

### ISSUE THREAD

**Issue by muyangren123321** [dl4to/dl4to#7]
Dear author, hello again. I am very grateful for developing this library, and I also appreciate your previous response to my question. I am now trying to implement the topology optimization of a cantilever beam under a point load based on your library. However, I have encountered some issues. The most important one is that after running the code and plotting the solution, the resulting structure still appears as the initial rectangular block, indicating that the topology optimization may not have been performed. Below is my code and the screenshot after running it.
# code
```
import torch
from dl4to.problem import Problem
from dl4to.solution import Solution
from dl4to.criteria import Compliance, VolumeConstraint
from dl4to.pde import FDM
from dl4to.topo_solvers import SIMP

E = 2.1e11  
v = 0.3     
sigma_ys = 2.5e8  
shape = [40, 10, 10]  
h = [0.01, 0.01, 0.01] 

F = torch.zeros(3, *shape)
F[-1, 39, 4, 9] = -1e5 
omiga_dirichlet = torch.zeros(3, *shape)
omiga_dirichlet[:, 0, :, :] = 1  
omiga_design = -torch.ones(1, *shape)
omiga_design[:, 0, :, :] = 1

problem = Problem(E, v, sigma_ys, h, omiga_dirichlet, omiga_design, F,
                  restrict_density_for_voxels_with_applied_forces=False)

problem.pde_solver = FDM()
camera_position = (0, 0.35, 0.2)
problem.plot(camera_position=camera_position,show_axislabels=True,show_ticklabels=True, display=True)

criterion = Compliance() + VolumeConstraint(max_volume_fraction=0.2, threshold_fct="relu")

simp = SIMP(criterion=criterion, binarizer_steepening_factor=1.03, n_iterations=100, lr=1e-2)
solution = simp(problem)
print(solution.get_θ(binary=True).mean())

camera_position = (0, 0.35, 0.2)
solution.plot(camera_position=camera_position, binary=False, solve_pde=True, display=True, use_pyvista=True)
```
# Locations of homogeneous Dirichlet boundary conditions 
![newplot (6)](https://github.com/user-attachments/assets/4aa95802-82c3-4704-af09-789a3a1964bd)
# Design space information
![newplot (7)](https://github.com/user-attachments/assets/736c5ec4-0d94-47b0-9f45-6ad22cd8bae0)
Furthermore, I have noticed that the density value of the element where the force is applied is not set to 1 in my Design Space Information plot.
![newplot (8)](https://github.com/user-attachments/assets/5341e16e-45df-437a-9675-56bbe5e18e4e)
# Force locations
![newplot (9)](https://github.com/user-attachments/assets/18b221ed-9bf5-4569-a653-60c228179297)
# Force directions
![newplot (10)](https://github.com/user-attachments/assets/c3d41db7-1c1e-4c13-b87f-1bb0fd12f685)
# Density distribution
![微信截图_20241021191104](https://github.com/user-attachments/assets/c1bbe164-15b2-4263-88a1-9b1f8f883ede)
# Normed displacements 
![微信截图_20241021191208](https://github.com/user-attachments/assets/f04f56a9-e7e1-4c2f-9693-44b141e3b0b3)
# Normalized von Mises stresses
![微信截图_20241021191307](https://github.com/user-attachments/assets/b792f817-7925-4ec4-81c2-2c1ae7445d39)
I suspect that there might be an issue with how I defined the F, Ω_dirichlet, and Ω_design variables. To investigate this, I saved the F, Ω_dirichlet, and Ω_design variables of the dataset[0] from the disc_complex dataset in your developed library as a CSV file for observation, as shown below.
# F
[tensor_data.csv](https://github.com/user-attachments/files/17459754/tensor_data.csv)
# Ω_dirichlet
[tensor_data1.csv](https://github.com/user-attachments/files/17459758/tensor_data1.csv)
# Ω_design
[tensor_data2.csv](https://github.com/user-attachments/files/17459761/tensor_data2.csv)
Upon examination, I noticed that in column 2712, which corresponds to the 2713th voxel grid, the applied forces in all three directions are set to 0, and no boundary conditions exist. However, in the design domain, its density value is set to 1, which means it is not a freely designable element. Therefore, my current question is whether for the design domain, only the elements with boundary conditions should be set to 1, while the others should be set to -1. If this is the case, then why isn't it the same in the dataset you provided? Therefore, I would like to ask how these variables should be assigned if one wants to design their own topology optimization problem.

I hope you can help me with this issue, and I am looking forward to your guidance. Thank you very much for your time and assistance.

**Comment by dl4to** [comment]
Hello muyangren123321, I recommend to never use forces or boundary conditions that are only defined on a single voxel. This can lead to problems with the finite differences PDE solver. Also, your force is too weak and as a result SIMP is working badly because the force leads to very stresses (so there is no further material needed and the design domain stays mostly empty). For instance, try using `F[-1, -2:, 3:6, -2:] = -5e9`.

This is the script that I ran and which works on my computer:

`import torch
from dl4to.problem import Problem
from dl4to.solution import Solution
from dl4to.criteria import Compliance, VolumeConstraint
from dl4to.pde import FDM
from dl4to.topo_solvers import SIMP

E = 2.1e11
ν = 0.3
σ_ys = 2.5e8
shape = [40, 10, 10]
h = [0.01, 0.01, 0.01]

F = torch.zeros(3, *shape)
F[-1, -2:, 3:6, -2:] = -5e9
Ω_dirichlet = torch.zeros(3, *shape)
Ω_dirichlet[:, 0, :, :] = 1
Ω_design = -torch.ones(1, *shape)
Ω_design[:, 0, :, :] = 1

problem = Problem(E=E, ν=ν, σ_ys=σ_ys, h=h,
                  Ω_dirichlet=Ω_dirichlet, Ω_design=Ω_design, F=F,
                  restrict_density_for_voxels_with_applied_forces=False)

problem.pde_solver = FDM()
criterion = Compliance() + VolumeConstraint(max_volume_fraction=0.2, threshold_fct="relu")

simp = SIMP(criterion=criterion, binarizer_steepening_factor=1.02, n_iterations=100, lr=1e-2)
solution = simp(problem)
print(solution.get_θ(binary=True).mean())

camera_position = (0, 0.35, 0.2)
solution.plot(solve_pde=True, binary=True, threshold=0.5)`

**Comment by muyangren123321** [comment]
Thank you very much for your response. I appreciate your perspective and I'm glad I could help. Have a great day and take care!
